adapters/from_sql.py

```diff
@@ -0,0 +1,194 @@
+"""Converts the data from SQL to expected format."""
+import json
+from data.release import Release
+from data.task import Task
+from adapters.to_sql import to_sql_tasks
+from adapters.to_sql import to_sql_xcom
+from resources.to_timestamp import to_timestamp
+from data.state import State
+from collections import namedtuple
+import logging
+
+STATE_FROM_STRING = {'none': State.UNUSED_STATUS,
+                     'running': State.PENDING,
+                     'success': State.FINISHED,
+                     'failed': State.FAILED,
+                     'shutdown': State.ABANDONED,
+                     'upstream_failed': State.PENDING}
+
+def read_releases(release_data, airflow_db):
+  """Reads in releases in the SQL tuple format, transforms them into objects.
+
+  Args:
+    release_data: the raw release data from SQL (tuple)
+    airflow_db: the airflow database connection object
+
+  Returns:
+    release_objects
+  """
+  release_named_tuple = namedtuple('Row', ['dag_id', 'execution_date'])
+  release_objects = {}  # resources.py expects a dict of objects
+  for item in release_data:  # iterate through each release in release_data
+    item = release_named_tuple._make(item)
+    release = Release()  # initialize the release object
+    started = to_timestamp(item.execution_date)
+    task_ids, most_recent_task, state = get_task_info(item.dag_id, started, airflow_db)
+    xcom_dict, green_sha = get_xcom(started, item.dag_id, airflow_db)
+    release.release_id = item.dag_id + '@' + str(item.execution_date)
+    release.tasks = task_ids
+    release.started = started
+    release.links = construct_links(green_sha)  # TODO(dommarques) these need to be implemented into airflow first, or we make our own way to get the links pylint: disable=line-too-long
+    release.labels = [item.dag_id]
+    release.state = state
+    release.branch, release.release_type = parse_dag_id(item.dag_id)
+    if xcom_dict is None:
```

Comment:
is this possible ? don't you need a XCOM to start a release @rkpagadala ?

Reply:
I didn't think it was possible either, but I kept running into this problem with certain releases, not really sure why but this was the only fix. They were the same releases where the tasks didn't exist for the release yet.

Reply:
Can we ignore those release in that case ? ie do not save or return them.

Reply:
Sure, that should be fairly simple